fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,220 @@
 
 
 class FendaClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        temperature: float = 0.5,
+        perFCL_loss_weights: List[float] = [0.0, 0.0],
+        cos_sim_loss_weight: float = 0.0,
+        contrastive_loss_weight: float = 0.0,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.perFCL_loss_weights = perFCL_loss_weights
+        self.cos_sim_loss_weight = cos_sim_loss_weight
+        self.contrastive_loss_weight = contrastive_loss_weight
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.temperature: float = temperature
+
+        # Need to save previous local module, global module and aggregated global module at each communication round
+        # to compute contrastive loss.
+        self.old_local_module: torch.nn.Module
+        self.old_global_module: torch.nn.Module
+        self.aggregated_global_module: torch.nn.Module
+
     def get_parameter_exchanger(self, config: Config) -> ParameterExchanger:
         assert isinstance(self.model, FendaModel)
         return FixedLayerExchanger(self.model.layers_to_exchange())
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+
+        preds = self.model(input)
+
+        preds["old_local_features"] = self.old_local_module.forward(input).view(len(preds["local_features"]), -1)
+        preds["old_global_features"] = self.old_global_module.forward(input).view(len(preds["global_features"]), -1)
+        if self.perFCL_loss_weights[0] != 0.0 or self.perFCL_loss_weights[1] != 0.0:
+            preds["aggregated_global_features"] = self.aggregated_global_module.forward(input).view(
+                len(preds["global_features"]), -1
+            )
+
+        if isinstance(preds, dict):
+            return preds
+        elif isinstance(preds, torch.Tensor):
+            return {"prediction": preds}
+        else:
+            raise ValueError("Model forward did not return a tensor or dictionary or tensors")
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+
+        # Save the parameters of the old model
+        assert isinstance(self.model.local_module, torch.nn.Module)
+        assert isinstance(self.model.global_module, torch.nn.Module)
+        self.old_local_module = copy.deepcopy(self.model.local_module)
+        self.old_global_module = copy.deepcopy(self.model.global_module)
```

Comment:
For the first round of federated learning, these weights are going to be random aren't they? If that's true I don't know if they will be useful for contrastive learning.

Reply:
Yeah that is true, I ran all of my experiments with fedavg warmup, maybe we can add a comment to point this out and refer them to use warm up module if they don't want them to be random?
Or one other option would be for the first round we shut down and possible contrastive loss and then activate it?

Reply:
Did you get a sense for what they do in the paper? If they had an approach to this, we can go with theirs. Otherwise, I'd say the later makes, where we only use previous client model comparison losses after the first training pass. 

It's a little tricky because some clients may not participate in every round and so their first round of client training may not necessarily be the first `server_round`. However, I think this would easily be accomplished by just moving the storage of these modules to the `get_parameters` function and any time we go to apply the loss, we check if the `old_local_module` and `old_global_module` values have been set. If they haven't then the loss is skipped.

Reply:
I don't get why it would be different if we move the module storage to get parameters function. For me it is more straight forward to store them when we set new parameters to our model.

Reply:
So the paper does not provide their code and they have not mentioned anything specific about this in the text.

Reply:
Of course they didn't 😂. Well let's go with your suggestion of no contrastive loss in the first round, as I think that makes the most sense.